First commit for project

ashleycoxley opened the pull request:
@harperj

ashleycoxley committed:
First commit for project

harperj commented on user_signup/app.yaml:
```diff
@@ -0,0 +1,11 @@
+
+libraries:
+- name: jinja2
+  version: latest
```
Generally preferred to include newlines at the end of files.  Not a big deal though.

harperj commented on user_signup/main.py:
```diff
@@ -0,0 +1,86 @@
+
+USER_RE = re.compile(r"^[a-zA-Z0-9_-]{3,20}$")
+PW_RE = re.compile(r"^.{3,20}$")
+EMAIL_RE = re.compile(r"^[\S]+@[\S]+.[\S]+$")
```
Did you write these regular expressions?  Does this regex correctly validate all possible emails?

ashleycoxley replied:
Didn't write them - they were given so we could focus on the other parts of this. I'll use better ones in the final blog project.

harperj replied:
👍  that seems fine

harperj commented on user_signup/main.py:
```diff
@@ -0,0 +1,86 @@
+		verify = self.request.get('verify')
+		email = self.request.get('email')
+
+		username_valid = validate_username(username)
```
I'd go ahead and turn this into a method `form_valid`:

``` python
form_is_valid = form_valid(
  username = self.request.get('username')
  password = self.request.get('password')
  verify = self.request.get('verify')
  email = self.request.get('email')
)

if form_is_valid:
  ...
```

harperj commented on user_signup/signup.html:
```diff
@@ -0,0 +1,47 @@
+<html>
+<head>
+	<title>Sign Up</title>
+	<link rel="stylesheet" href="style.css">
```
👍

harperj commented on user_signup/signup.html:
```diff
@@ -0,0 +1,47 @@
+<body>
+	<h2>Signup</h2>
+	<form method='post'>
+		<table>
```
Why'd you use a table here?  Usually, you'll only want to use the `<table>` tag when you're actually laying out data in a table, rather than to specify a layout.

ashleycoxley replied:
This, also, was just copied from the course - I think they were trying to get around having to use CSS for the layout. Thanks for the tip. I'll revamp it in the final version of the blog.

harperj replied:
👍  that seems fine

harperj reviewed (changes requested):
Some changes to / answers about your code would be good here.

harperj commented:
Ok to the point of what I was saying earlier, just push changes to `user_signup`.  You'll just checkout the branch, make whatever changes you like, then either commit them the usual way with a useful message or do: `git commit --fixup=HEAD`.  This tells git that the commit is just meant to "fix up" the last commit (at `HEAD`).

ashleycoxley committed:
fixup! First commit for project

harperj commented on user_signup/main.py:
```diff
@@ -7,18 +7,72 @@
+EMAIL_ERROR = "Invalid email address"
+
+
+def build_validation_re(USER_RE, PW_RE, EMAIL_RE):
```
This and `build_error_messages` don't really need to be functions, you can just have it look like:

``` python
VALIDATION_MATCHERS = {
  'username': re.compile(USER_RE),
  'password': re.compile(PW_RE),
  'email': re.compile(EMAIL_RE)
}
```

harperj commented on user_signup/main.py:
```diff
@@ -7,18 +7,72 @@
+def validate(form_input, input_type):
+	validation_re = build_validation_re(USER_RE, PW_RE, EMAIL_RE)
+	validation_re = validation_re[input_type]
+	return validation_re.match(form_input)
```
As I mentioned above, then this could be just:

``` python
def validate(form_input, input_type):
  validation_re = VALIDATION_MATCHERS[input_type]
  return validation_re.match(form_input)
```

ashleycoxley replied:
👍

harperj commented on user_signup/main.py:
```diff
@@ -7,18 +7,72 @@
+		email_valid = validate(email, 'email')
+	
+	form_valid = True
+	template_values = {'username': '',
```
Do you really have to pre-specify this?

I bet you can just do:

``` python
template_values = {}
```

ashleycoxley replied:
Unfortunately, yes, because I need the default values to be empty strings.

ashleycoxley committed:
Added user DB storage

 - Created User datastore model
 - Database entries for each user upon POST
 - Added a new input verification: username must not exist already
 - Added password and user_id cookie hashing functions
 - Set cookie for user_id and used datastore to return username in the 'success' page
 - Changed variable dictionaries from functions to global variables per Harper's last review

harperj commented:
You should probably merge this now..
